### ISSUE TITLE
What happens if the container already has the interface "inside the container"?

### ISSUE THREAD

**Issue by aanm** [appc/cni#112]
> Name of the interface inside the container. This is the name that should be assigned to the interface created inside the container (network namespace); consequently it must comply with the standard Linux restrictions on interface names.

The spec doesn't have anything related to that. So what is the procedure? Are the plugins allowed to remove, make the necessary modifications, and then recreate the interface with the same name?

**Comment by dcbw** [comment]
Removing the interface may break existing connections and cause apps inside the container to fail when their interface goes away and sockets are closed and things like that.  So I would imagine that completely removing the interface would not be desirable.

I'm assuming this item is in the spec because apps may reference the network interface by name or something, or maybe the runtime may need to modify the interface in some way.

It's worth noting that the libnetwork/CNM API doesn't pass an expected interface name into the plugin, but does receive the interface name back from the plugin.  This would let the plugin itself choose whether to create additional interfaces in the namespace, or to set up the existing interface for the new network.  The interface name is then passed up to the runtime.

So I guess I'm wondering why CNI sends the expected interface down to the plugins at all?

**Comment by aanm** [comment]
>It's worth noting that the libnetwork/CNM API doesn't pass an expected interface name into the plugin, but does receive the interface name back from the plugin. This would let the plugin itself choose whether to create additional interfaces in the namespace, or to set up the existing interface for the new network. The interface name is then passed up to the runtime.

Accordingly with the spec the only thing passed up to the runtime is the IPs assigned and the DNS info.
>Result:
IPs assigned to the interface. This is either an IPv4 address, an IPv6 address, or both.
DNS information. Dictionary that includes DNS information for nameservers, domain, search domains and options.

**Comment by dcbw** [comment]
Yeah, I think that should get changed.  The runtime should pass back the interface name and all that, instead of accepting an interface name from the runtime.  Otherwise I can't see how multiple networks in a container can work for some cases, even with the same plugin.

**Comment by steveeJ** [comment]
@dcbw 
> The runtime should pass back the interface name and all that, instead of accepting an interface name from the runtime.
I would rather not guess which "runtime" to replace for "plugin". If I had to I'd say the first one?

**Comment by steveeJ** [comment]
Ping @dcbw

**Comment by dcbw** [comment]
@steveeJ I'm working on a PR to pass the interface name back to the runtime.  It doesn't yet re-try interface creation with different names, but at least it gets the API ready.
